### PR TITLE
[Misc] Support routing logic simulation

### PR DESCRIPTION
**Pull request by minosfuture**
## Purpose

This is to support token-to-expert routing logic customization. We need this to work around imbalanced token-to-expert selection and support balanced selection for benchmark purpose.

Customization can be implemented by extending RoutingStrategy and implement route_tokens, being registered with RoutingSimulator, and selected by VLLM_MOE_ROUTING_STRATEGY envvar.

Thanks @tlrmchlsmth for initial implementation and handover!

Next PR:
* Add metrics to confirm effectiveness of customized routing
* Potentially extend RoutingSimulator into a general Router class and modularize FusedMoE.select_experts implementation.

## Test Plan

pytest tests/test_routing_simulator.py

## Test Result

passed

### CODE REVIEW

File: vllm/envs.py [hunk 1]
```diff
@@ -963,6 +963,14 @@ def get_vllm_port() -> Optional[int]:
     "VLLM_MAX_TOKENS_PER_EXPERT_FP4_MOE":
     lambda: int(os.getenv("VLLM_MAX_TOKENS_PER_EXPERT_FP4_MOE", "163840")),
 
+    # MoE routing strategy selector. Available strategies:
+    # - "softmax": Standard softmax-based routing
+    # - "uniform_random": Uniform random routing for perfect load balancing
+    # - "weighted_random": Weighted random routing with temperature scaling
+    # Note: custom strategies may not produce correct model outputs
+    "VLLM_MOE_ROUTING_STRATEGY":
+    lambda: os.environ.get("VLLM_MOE_ROUTING_STRATEGY", "").lower(),
```
Comment (tlrmchlsmth) [inline vllm/envs.py]: Any plans for routing strategies besides `softmax` that would result in correct outputs? If not, let's give it a name that indicates that it should be used for testing? E.g. `VLLM_MOE_ROUTING_SIMULATOR`

File: vllm/envs.py [hunk 2]
```diff
@@ -963,6 +963,14 @@ def get_vllm_port() -> Optional[int]:
     "VLLM_MAX_TOKENS_PER_EXPERT_FP4_MOE":
     lambda: int(os.getenv("VLLM_MAX_TOKENS_PER_EXPERT_FP4_MOE", "163840")),
 
+    # MoE routing strategy selector. Available strategies:
+    # - "softmax": Standard softmax-based routing
```
Comment (tlrmchlsmth) [inline vllm/envs.py]: looks like we don't handle `VLLM_MOE_ROUTING_STRATEGY` being equal to `softmax` anywhere

File: vllm/model_executor/layers/fused_moe/routing_simulator.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+            high=global_num_experts,
+            size=(num_tokens, top_k),
+            dtype=indices_type,
+            device=hidden_states.device,
+        )
+
+        # All-ones weights
+        topk_weights = torch.ones(
+            (num_tokens, top_k),
+            dtype=torch.float32,
```
Comment (tlrmchlsmth) [inline vllm/model_executor/layers/fused_moe/routing_simulator.py]: Are these always fp32? (I think so but not sure)

File: vllm/model_executor/layers/fused_moe/routing_simulator.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+
+        # Get corresponding weights
+        topk_weights = routing_probs.gather(dim=-1, index=topk_ids)
+
+        # Renormalize weights
+        topk_weights = topk_weights / topk_weights.sum(dim=-1, keepdim=True)
+
+        # Convert indices to requested type
+        if indices_type is not None:
+            topk_ids = topk_ids.to(dtype=indices_type)
```
Comment (tlrmchlsmth) [inline vllm/model_executor/layers/fused_moe/routing_simulator.py]: The `routed_logits` will still be bogus in the case of `--load-format=dummy` -- Is this strategy usable in this case?

If not, what is the expected use case for `WeightedRandomRouting`?

File: vllm/model_executor/layers/fused_moe/routing_simulator.py (hunk 3 above)
Comment (minosfuture) [reply vllm/model_executor/layers/fused_moe/routing_simulator.py]: yep, it is. Confirmed with grouped_topk.

File: vllm/envs.py (hunk 2 above)
Comment (minosfuture) [reply vllm/envs.py]: Removed. Thanks for the catch!

File: vllm/envs.py (hunk 1 above)
Comment (minosfuture) [reply vllm/envs.py]: I'm thinking of extending RoutingSimulator to a general class `Router` in next PRs and break FusedMoE.select_experts into multiple RoutingStrategy subclasses. wdyt?

File: vllm/model_executor/layers/fused_moe/routing_simulator.py (hunk 4 above)
Comment (minosfuture) [reply vllm/model_executor/layers/fused_moe/routing_simulator.py]: WeightedRandomRouting dials down the imbalance according to the temperature param. So it can keep the original distribution, but make it less imbalanced.

File: .gitignore [hunk 5]
```diff
@@ -203,3 +203,5 @@ shellcheck*/
 
 # Ignore moe/marlin_moe gen code
 csrc/moe/marlin_moe_wna16/kernel_*
+
+nixl_installer/
```
Comment (tlrmchlsmth) [inline .gitignore]: Is this an accident? It's unrelated to this PR and I'm not sure what creates this dir

File: vllm/envs.py (hunk 1 above)
Comment (tlrmchlsmth) [reply vllm/envs.py]: Could you elaborate a bit more on what the specific RoutingStrategy subclasses will be and why there should be an environment variable to select between them? 

Are you planning on adding different strategies that have different performance characteristics?

File: .gitignore (hunk 5 above)
Comment (minosfuture) [reply .gitignore]: yes, I should remove it. It's created by tools/install_nixl.sh for all2all.

File: vllm/envs.py (hunk 1 above)
Comment (minosfuture) [reply vllm/envs.py]: We can organize strategies into the following
1. common strategies provided by vllm
  * GroupedTopk
  * FusedTopk
2. simulated strategies for benchmark
  * UniformRandomRouting
  * WeightedRandomRouting
  * ImbalancedRouting (to be added)
  * and yes, we can add more strategies here for different performance characteristics
3. custom strategies registered by model implementation
  // this can replace the current custom_routing_function

strategy selection can be finalized in FusedMoE init function, consolidating this envvar VLLM_MOE_ROUTING_STRATEGY and model configuration.
And VLLM_MOE_ROUTING_STRATEGY can be used to override model config, e.g., for benchmark purpose.

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 6]
```diff
@@ -1275,6 +1304,16 @@ def select_experts(
...
+        # Check if we should use a routing simulation strategy
+        routing_strategy = envs.VLLM_MOE_ROUTING_STRATEGY
+        if routing_strategy != "":
+            return FusedMoE.select_experts_with_simulated_strategy(
+                hidden_states=hidden_states,
+                router_logits=router_logits,
+                top_k=top_k,
+                strategy=routing_strategy,
+                indices_type=indices_type)
+
```
Comment (wpc) [inline vllm/model_executor/layers/fused_moe/layer.py:1289]: seems we can inline this call?

File: vllm/model_executor/layers/fused_moe/routing_simulator.py [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+            top_k: Number of experts to select per token
+            indices_type: Data type for expert indices
+
+        Returns:
+            tuple of (topk_weights, topk_ids)
+        """
+        pass
+
+
+class UniformRandomRouting(RoutingStrategy):
```
Comment (wpc) [inline vllm/model_executor/layers/fused_moe/routing_simulator.py]: this is just a special case for imbalance routing? maybe user can specify a function name to generate dispatching distribution instead of having multiple subclasses?

File: vllm/model_executor/layers/fused_moe/routing_simulator.py [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+
+
+class WeightedRandomRouting(RoutingStrategy):
+    """
+    Weighted random routing strategy.
+
+    This strategy samples experts randomly based on the router logits
+    probabilities, providing a stochastic routing approach that still respects
+    the learned routing preferences while introducing randomness for load
+    balancing.
```
Comment (wpc) [inline vllm/model_executor/layers/fused_moe/routing_simulator.py]: can you explain more the usecase for this strategy?

File: vllm/model_executor/layers/fused_moe/routing_simulator.py (hunk 4 above)
Comment (minosfuture) [reply vllm/model_executor/layers/fused_moe/routing_simulator.py]: removed WeightedRandomRouting on a second thought: we would be more interested in certain distribution or enforcing a specified imbalanced factor than keeping the original distribution.

File: vllm/model_executor/layers/fused_moe/layer.py (hunk 6 above)
Comment (minosfuture) [reply vllm/model_executor/layers/fused_moe/layer.py:1289]: updated

File: vllm/model_executor/layers/fused_moe/routing_simulator.py (hunk 8 above)
Comment (minosfuture) [reply vllm/model_executor/layers/fused_moe/routing_simulator.py]: removed WeightedRandomRouting on a second thought: we would be more interested in certain distribution or enforcing a specified imbalanced factor than keeping the original distribution.

File: vllm/model_executor/layers/fused_moe/routing_simulator.py (hunk 7 above)
Comment (minosfuture) [reply vllm/model_executor/layers/fused_moe/routing_simulator.py]: great point. Updated. I'll also add the imbalanced factor implementation in a followup PR.

File: vllm/envs.py (hunk 1 above)
Comment (wpc) [reply vllm/envs.py]: name "VLLM_MOE_ROUTING_STRATEGY" feels misleading. I feel moe routing strategy is referring token choice or expert choice or threshold based etc. Maybe we can rename it to something like VLLM_MOE_ROUTING_SIMULATION_STRATEGY?

Comment (minosfuture) [reply vllm/envs.py]: sounds like it's preferred to keep this envvar solely for simulation!
updated and changed the naming. thx!

**Review by tlrmchlsmth — APPROVED** [review]

**Comment by minosfuture** [comment]
@tlrmchlsmth  the failed test in CI is passed locally. I wonder if this test is flaky. Should we bypass?

**Comment by DarkLight1337** [comment]
Merging
